harness/determined/pytorch/_pytorch_context.py

```diff
@@ -57,4 +102,359 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        """
+        check.false(self._use_amp, "Please only call configure_apex_amp once.")
+        if self.hvd_config.use:
+            check.eq(
+                num_losses,
+                1,
+                "When using parallel/distributed training, "
+                "Determined only supports configure_apex_amp with num_losses = 1",
+            )
+
+        self._use_amp = True
+
+        if self.hvd_config.use:
+            check.eq(
+                self.hvd_config.aggregation_frequency,
+                1,
+                "Mixed precision training (AMP) is not supported with "
+                "aggregation frequency > 1.",
+            )
+
+        check.true(
+            torch.cuda.is_available(),
+            "Mixed precision training (AMP) is supported only on GPU slots.",
+        )
+
+        logging.info(f"Enabling mixed precision training with opt_level: {opt_level}.")
+        models, optimizers = apex.amp.initialize(
+            models=models,
+            optimizers=optimizers,
+            enabled=enabled,
+            opt_level=opt_level,
+            cast_model_type=cast_model_type,
+            patch_torch_functions=patch_torch_functions,
+            keep_batchnorm_fp32=keep_batchnorm_fp32,
+            master_weights=master_weights,
+            loss_scale=loss_scale,
+            cast_model_outputs=cast_model_outputs,
+            num_losses=num_losses,
+            min_loss_scale=min_loss_scale,
+            max_loss_scale=max_loss_scale,
+            verbosity=verbosity
+            if self.distributed.get_rank() == 0 or self.env.experiment_config.debug_enabled()
+            else 0,
+        )
+        if not isinstance(models, list):
+            self.models = [models]
+        if not isinstance(optimizers, list):
+            self.optimizers = [optimizers]
+        return models, optimizers
+
+    def _should_communicate_and_update(self) -> bool:
+        if self._current_batch_idx is None:
+            raise det.errors.InternalException("Training hasn't started.")
+        return (self._current_batch_idx + 1) % self.hvd_config.aggregation_frequency == 0
+
+    def _backward(
+        self,
+        loss: torch.Tensor,
+        gradient: Optional[torch.Tensor] = None,
+        retain_graph: bool = False,
+        create_graph: bool = False,
+    ) -> None:
+        """Compute the gradient of current tensor w.r.t. graph leaves.
+
+        The arguments are used in the same way as ``torch.Tensor.backward``.
+        See https://pytorch.org/docs/1.4.0/_modules/torch/tensor.html#Tensor.backward for details.
+
+        .. warning::
+
+            When using distributed training, we don't support manual gradient accumulation.
+            That means the gradient on each parameter can only be calculated once on each batch.
+            If a parameter is associated with multiple losses, you can only call backward on
+            these losses for no more than once. Or you could set require_grads flag of a
```

Comment:
"If a parameter is associated with multiple losses, you can either choose to call backward on only one of those losses, or you could set the ``require_grads`` flag ..."

"However you could do gradient accumulation across batches by ..."

(the second suggestion is courtesy of @aaron276h)